Add ngettext() feature.

cvwillegen opened the pull request:
Some time ago, the gettext() function was made available to be used from script files to provide localized text for plugins. Since languages can have rules to create plural forms from singular sentences, an ngettext() function should be provided as well.

cvwillegen committed:
Add ngettext() feature.

cvwillegen committed:
Add missing files to Filelist.

cvwillegen committed:
Add type cast to function call.

cvwillegen committed:
Also fix type cast in the else branch...

cvwillegen committed:
Fix expected message in cp1251.

cvwillegen committed:
Correct link in the documentation.

trofi committed:
patch 9.1.1055: make install fails because of a missing dependency

Problem:  make install fails because of a missing dependency
Solution: add explicit dependencies for tutor/{en,it}
          (Sergei Trofimovich)

Without the change `make install -j16` fails
sometimes due to race condition to create a directory.

It's best reproducible with `make --shuffle`:

    $ make install --shuffle
    ...
    bash install-sh -c -d /vim-9.1.0990/share/vim/vim91/tutor
    chmod 755 /vim-9.1.0990/share/vim/vim91/tutor
    cp ../runtime/tutor/README* ../runtime/tutor/tutor* /vim-9.1.0990/share/vim/vim91/tutor
    cp ../runtime/tutor/en/* /vim-9.1.0990/share/vim/vim91/tutor/en/
    cp: target '/vim-9.1.0990/share/vim/vim91/tutor/en/': No such file or directory
    make[1]: [Makefile:2487: installtutor] Error 1 (ignored) shuffle=2340321974
    cp ../runtime/tutor/it/* /vim-9.1.0990/share/vim/vim91/tutor/it/
    cp: target '/vim-9.1.0990/share/vim/vim91/tutor/it/': No such file or directory
    make[1]: [Makefile:2488: installtutor] Error 1 (ignored) shuffle=2340321974

The fix adds a dependency on targeted directories used to install tutor.

Before the change it too 2-3 attempts to reproduce the install failure.
After the change `vim` survives 150 install attempts without failures.

closes: #16531

Signed-off-by: Sergei Trofimovich <slyich@gmail.com>
Signed-off-by: Christian Brabandt <cb@256bit.org>

saccarosium committed:
runtime(hyprlang): fix string recognition

fixes: #16064
closes: #16527

Signed-off-by: Luca Saccarola <github.e41mv@aleeas.com>
Signed-off-by: Christian Brabandt <cb@256bit.org>

saccarosium committed:
runtime(dist/vim9): fix regressions in dist#vim9#Open

fixes: #16533
fixes: #16532
closes: #16535

Signed-off-by: Luca Saccarola <github.e41mv@aleeas.com>
Signed-off-by: Christian Brabandt <cb@256bit.org>

saccarosium committed:
runtime(netrw): upstream snapshot of v176

relevant commits:
- deprecate!: netrw#Launch, netrw#Open and gx mappings
- refactor: move some utility functions in a private file
- feat: add function to deprecate features
- refactor!: remove NetrwClean command and function
- refactor: use appropriate directories to store temporary files
- refactor: better way to call vim.ui.open
- refactor(Open): prefer lua wrapper function instead of cmdline
- refactor!: drop vim 7 checks
- refactor: use vim.ui.open when using neovim
- refactor: remove s:CheckIfKde
- refactor: balloon functionality
- refactor!: remove netrw#Access function

closes: #16519

Signed-off-by: Luca Saccarola <github.e41mv@aleeas.com>
Signed-off-by: Christian Brabandt <cb@256bit.org>

glepnir committed:
patch 9.1.1056: Vim doesn't highlight to be inserted text when completing

Problem:  Vim doesn't highlight to be inserted text when completing
Solution: Add support for the "preinsert" 'completeopt' value
          (glepnir)

Support automatically inserting the currently selected candidate word
that does not belong to the latter part of the leader.

fixes: #3433
closes: #16403

Signed-off-by: glepnir <glephunter@gmail.com>
Signed-off-by: Christian Brabandt <cb@256bit.org>

RestorerZ commented on src/testdir/ru_RU/LC_MESSAGES/__PACKAGE__.po:
```diff
@@ -0,0 +1,33 @@
+msgid_plural "%d buffers unloaded"
+msgstr[0] "%d буфер удалён из памяти for __PACKAGE__"
+msgstr[1] "%d буфера удалено из памяти for __PACKAGE__"
+msgstr[2] "%d буферов удалено из памяти for __PACKAGE__"
```
~There's something wrong with the text encoding here.~
~Compare it to this commit~
https://github.com/vim/vim/pull/16529/commits/dab369a90f270b01be8e7fb85b70083538056686#diff-7577c056033f5d21b81472320cd084774480c4c6aa77a1a32f55181460397ab1
~It looks like Windosw-1251 encoding was just saved to UTF-8 without converting it.~

RestorerZ replied:
Checked it again, it's fine. There must be some glitch in Chromium

RestorerZ commented on src/testdir/test_gettext_utf8.vim:
```diff
@@ -14,14 +14,28 @@ func Test_gettext()
+
+      call assert_equal('%d буфер удалён из памяти for __PACKAGE__', ngettext("%d buffer unloaded", "%d buffers unloaded", 1, "__PACKAGE__"))
+      call assert_equal('%d буфера удалено из памяти for __PACKAGE__', ngettext("%d buffer unloaded", "%d buffers unloaded", 2, "__PACKAGE__"))
+      call assert_equal('%d буферов удалено из памяти for __PACKAGE__', ngettext("%d buffer unloaded", "%d buffers unloaded", 5, "__PACKAGE__"))
```
~Same here.~
https://github.com/vim/vim/pull/16529/commits/dab369a90f270b01be8e7fb85b70083538056686#diff-f1f42e08d5b4e0d62dec4c8994afcf28757c584e8daa69eaa579cc80aa9965df

RestorerZ replied:
Checked it again, it's fine. There must be some glitch in Chromium

chrisbra committed:
runtime(netrw): correct wrong version check

The version check introduced with commit edd4ac3e8 has a logic error.
Also it should only trigger when trying to use the netrw plugin and not
always.

fixes: #16541

Signed-off-by: Christian Brabandt <cb@256bit.org>

zeertzjq committed:
patch 9.1.1057: Superfluous cleanup steps in test_ins_complete.vim

Problem:  Superfluous cleanup steps in test_ins_complete.vim.
Solution: Remove unnecessary :bw! and :autocmd! commands.
          Also remove unnecessary STRLEN() in insexpand.c
          (zeertzjq)

closes: #16542

Signed-off-by: zeertzjq <zeertzjq@outlook.com>
Signed-off-by: Christian Brabandt <cb@256bit.org>
